File: plugins/plugin-nextcloud-talk/rust/src/actions/send_message.rs
use async_trait::async_trait;
use serde_json::Value;

use super::{ActionContext, NextcloudTalkAction};
use crate::error::Result;

/// Action that builds a JSON payload for sending a Nextcloud Talk message.
pub struct SendMessageAction;

#[async_trait]
impl NextcloudTalkAction for SendMessageAction {
    fn name(&self) -> &'static str {
        "SEND_NEXTCLOUD_TALK_MESSAGE"
    }

    fn description(&self) -> &'static str {
        "Send a message to a Nextcloud Talk room"
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context.message.get("source").and_then(|v| v.as_str());
        Ok(source == Some("nextcloud-talk"))
    }

    async fn execute(&self, context: &ActionContext) -> Result<Value> {
        let response_text = context
            .state
            .get("response")
            .and_then(|v| v.get("text"))
            .and_then(|v| v.as_str())
            .unwrap_or("");

        Ok(serde_json::json!({
            "action": self.name(),
            "room_token": context.room_token,
            "text": response_text,
            "reply_to_message_id": context.message.get("message_id")
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_validate_nextcloud_talk_source() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "nextcloud-talk",
                "text": "Hello"
            }),
            room_token: "abc123".to_string(),
            user_id: "user1".to_string(),
            state: serde_json::json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_non_nextcloud_talk_source() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "telegram",
                "text": "Hello"
            }),
            room_token: "abc123".to_string(),
            user_id: "user1".to_string(),
            state: serde_json::json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_execute() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "nextcloud-talk",
                "message_id": "12345"
            }),
            room_token: "room123".to_string(),
            user_id: "user1".to_string(),
            state: serde_json::json!({
                "response": {
                    "text": "Hello, world!"
                }
            }),
        };

        let result = action.execute(&context).await.unwrap();
        assert_eq!(result["action"], "SEND_NEXTCLOUD_TALK_MESSAGE");
        assert_eq!(result["room_token"], "room123");
        assert_eq!(result["text"], "Hello, world!");
    }
}

File: plugins/plugin-nextcloud-talk/typescript/src/actions/sendMessage.ts
import type {
  Action,
  ActionExample,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { NEXTCLOUD_TALK_SERVICE_NAME } from "../constants";
import type { NextcloudTalkService } from "../service";

export const SEND_MESSAGE_ACTION = "SEND_NEXTCLOUD_TALK_MESSAGE";

export const sendMessageAction: Action = {
  name: SEND_MESSAGE_ACTION,
  similes: [
    "NEXTCLOUD_TALK_SEND_MESSAGE",
    "NEXTCLOUD_TALK_REPLY",
    "NEXTCLOUD_TALK_MESSAGE",
    "NC_TALK_SEND",
    "NC_TALK_REPLY",
  ],
  description: "Send a message to a Nextcloud Talk room",

  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const source = message.content?.source;
    return source === "nextcloud-talk";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    const service = (await runtime.getService(NEXTCLOUD_TALK_SERVICE_NAME)) as
      | NextcloudTalkService
      | undefined;

    if (!service) {
      if (callback) {
        await callback({
          text: "Nextcloud Talk service not available",
        });
      }
      return { success: false, error: "Nextcloud Talk service not initialized" };
    }

    const currentState = state ?? (await runtime.composeState(message));
    const responseText = currentState.values?.response?.toString() || "";
    const roomToken = message.content?.roomToken as string | undefined;

    if (!roomToken) {
      if (callback) {
        await callback({
          text: "No room token available",
        });
      }
      return { success: false, error: "Missing room token" };
    }

    if (callback) {
      await callback({
        text: responseText,
        action: SEND_MESSAGE_ACTION,
      });
    }

    return {
      success: true,
      data: {
        action: SEND_MESSAGE_ACTION,
        roomToken,
        text: responseText,
        replyToMessageId: message.content?.messageId,
      },
    };
  },

  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "Send a message to this Nextcloud Talk room",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "I'll send a message to this room now.",
          actions: [SEND_MESSAGE_ACTION],
        },
      },
    ],
  ] as ActionExample[][],
};
